lib/serialization/compositeSpec.ts

```diff
@@ -8,29 +8,94 @@ export interface CompositeType {
   [key: string]: any;
 }
 
+export interface PolymorphicDerivedType {
+  /**
+   * A CompositeTypeSpec that "inherits" from this CompositeTypeSpec.
+   */
+  derivedTypeSpec: CompositeTypeSpec | string;
+
+  /**
+   * The value of the discriminator property that would indicate that this CompositeTypeSpec is the
+   * target TypeSpec for the value being serialized or deserialized.
+   */
+  discriminatorPropertyValue: any;
+}
+
+/**
+ * The properties that specify the derived types of a CompositeSpecType.
+ */
+export interface PolymorphicDerivedTypes {
+  /**
+   * The CompositeTypeSpecs that "inherit" from this CompositeTypeSpec.
+   */
+  derivedTypes: PolymorphicDerivedType[];
+
+  /**
+   * The name of the property that determines what the actual type the polymorphic object is.
+   */
+  discriminatorPropertyName: string;
+
+  /**
+   * The serialized name of the property that determines what the actual type of the polymorphic
+   * object is. If this is not provided, then the discriminatorPropertyName should be used.
+   */
+  discriminatorPropertySerializedName?: string;
+
+  /**
+   * The value of the discriminator property that would indicate that this CompositeTypeSpec is the
+   * target TypeSpec for the value being serialized or deserialized.
+   */
+  discriminatorPropertyValue: any;
+}
+
+/**
+ * The options that specify polymorphism for a CompositeTypeSpec.
+ */
+export interface Polymorphism {
+  /**
+   * The base types for this CompositeTypeSpec.
+   */
+  inheritsFrom?: (CompositeTypeSpec | string) | (CompositeTypeSpec | string)[];
+
+  /**
+   * The details of the derived types of this CompositeTypeSpec.
+   */
+  inheritedBy?: PolymorphicDerivedTypes;
+}
+
+/**
+ * The TypeSpec for serializing and deserializing a CompositeType.
+ */
 export interface CompositeTypeSpec extends TypeSpec<CompositeType, CompositeType> {
   /**
    * The name of the composite type (class) that this CompositeTypeSpec describes.
    */
   typeName: string;
 
+  /**
+   * The options that specify polymorphism for this CompositeTypeSpec.
+   */
+  polymorphism?: Polymorphism;
+
   /**
    * The specifications for each of the properties that exist on the type that this
    * CompositeTypeSpec describes.
    */
-  propertySpecs: { [propertyName: string]: PropertySpec };
+  propertySpecs?: { [propertyName: string]: PropertySpec };
 }
 
 /**
  * A type specification that describes how to validate and serialize a Composite value.
  */
-export function compositeSpec(typeName: string, propertySpecs: { [propertyName: string]: PropertySpec }): CompositeTypeSpec {
+export function compositeSpec(args: { typeName: string, polymorphism?: Polymorphism, propertySpecs?: { [propertyName: string]: PropertySpec } }): CompositeTypeSpec {
```

Comment:
Rather than defining the type inline, do you think defining an interface called CompositeSpecParameters would be better. That makes things cleaner and provides a better intellisense experience as someone can F12 on the type and find out its properties.

Reply:
nit: Instead of using args can you use parameters. Initially i got confused with args and javascript reserved word `arguments` for getting the arguments array of a function.

Reply:
Fixed in https://github.com/Azure/ms-rest-js/pull/63/commits/89261f3a845e1cd0ad505743b7cfc48428440af7.